django/contrib/postgres/constraints.py

```diff
@@ -0,0 +1,85 @@
+        if index_type and index_type.lower() not in {'gist', 'spgist'}:
+            raise ValueError('Exclusion constraints only support GiST or SP-GiST indexes.')
+        if not expressions:
+            raise ValueError('The expressions argument is mandatory.')
+        if condition is not None and not isinstance(condition, Q):
+            raise ValueError('The condition argument needs to be either None or a Q-instance.')
+        self.expressions = expressions
+        self.index_type = index_type
+        self.condition = condition
+        super().__init__(name=name)
+
+    def _parse_expressions(self, compiler, connection, query):
+        expressions = []
+        for expression, _ in self.expressions:
+            if isinstance(expression, (F, str)):
+                base = F(expression) if isinstance(expression, str) else expression
+                expression = base.resolve_expression(query=query, simple_col=True)
+            expression = expression.resolve_expression(query=query)
+            sql, params = expression.as_sql(compiler, connection)
+            expressions.append(sql % params)
+        return [
+            '%s WITH %s' % (expression, self.expressions[idx][1])
+            for idx, expression in enumerate(expressions)
+        ]
+
+    def _parse_where(self, compiler, schema_editor, query):
+        if self.condition is None:
+            return ''
+        where = query.build_where(self.condition)
+        sql, params = where.as_sql(compiler, schema_editor.connection)
+        params = tuple(schema_editor.quote_value(p) for p in params)
+        return ' WHERE ({})'.format(sql % params)
+
+    def constraint_sql(self, model, schema_editor):
+        query = Query(model)
+        compiler = schema_editor.connection.ops.compiler('SQLCompiler')(query, schema_editor.connection, 'default')
+        expressions = self._parse_expressions(compiler, schema_editor.connection, query)
+        where = self._parse_where(compiler, schema_editor, query)
+        return self.template % {
+            'name': schema_editor.quote_name(self.name),
+            'index_type': self.index_type or 'GIST',
+            'expressions': ', '.join(expressions),
+            'where': where,
+        }
+
+    def deconstruct(self):
+        kwargs = {
+            'expressions': self.expressions,
+            'name': self.name,
+        }
+        if self.condition is not None:
+            kwargs['condition'] = self.condition
+        if self.index_type is not None:
+            kwargs['index_type'] = self.index_type
+        return self.__class__.__qualname__, (), kwargs
+
+    def __eq__(self, other):
+        return (
+            isinstance(other, self.__class__) and
+            self.name == other.name and
+            self.index_type == other.index_type and
+            self.expressions == other.expressions and
+            self.condition == other.condition
+        )
```

Comment:
Do we need to define `__eq__`?

Reply:
Hmm. There's an `__eq__` in `CheckConstraint` and `UniqueConstraint`. I don't see the harm in having it.

Reply:
All tests work without it that's why I'm wondering if we need it.

Reply:
Maybe it's just because there's no test coverage. I didn't check the coverage report, but I'll do that before unchecking the "patch needs improvement".